File: Python/add_two_numbers.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        # 创建头节点
        result = curr = ListNode()
        # 保留进位
        remainder = 0
        while l1 or l2:
            x = l1.val if l1 else 0
            y = l2.val if l2 else 0
            total = x + y + remainder
            # 得到首节点 个位数
            curr.next = ListNode(total % 10)
            # 求出进位
            remainder = total // 10
            if l1:
                l1 = l1.next
            if l2:
                l2 = l2.next
            curr = curr.next
        # 如果有进位则传入
        if remainder:
            curr.next = ListNode(remainder)
        return result.next

File: Java/add_two_numbers.java

//Definition for singly-linked list.

class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}

class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode head = null, tail = null;
        int x, y, total;
        int carry = 0;
        while (l1 != null || l2 != null) {
            if (l1 != null) x = l1.val;
            else x = 0;
            if (l2 != null) y = l2.val;
            else y = 0;
            total = x + y + carry;
            carry = total / 10;
            if (head == null) {
                head = tail = new ListNode(total % 10);
            } else {
                tail.next = new ListNode(total % 10);
                tail = tail.next;
            }
            if (l1 != null) l1 = l1.next;
            if (l2 != null) l2 = l2.next;
        }
        if (carry > 0) tail.next = new ListNode(carry);
        return head;
    }
}

class main {
    public static void main(String[] args) {
        ListNode l1 = new ListNode(0);
        ListNode l2 = new ListNode(0);
        Solution solution = new Solution();
        ListNode head = solution.addTwoNumbers(l1, l2);
        System.out.println("val:" + head.val);
    }
}

